apps/src/templates/instructions/ReviewTab.jsx

```diff
@@ -1,18 +1,30 @@
 import React, {Component} from 'react';
-//import PropTypes from 'prop-types';
+import PropTypes from 'prop-types';
 import javalabMsg from '@cdo/javalab/locale';
 import Comment from './codeReview/Comment';
-import {demoComments} from './codeReview/commentShape';
+import {commentShape} from './codeReview/commentShape';
 import CommentEditor from './codeReview/CommentEditor';
 
 export default class ReviewTab extends Component {
-  // Once we have real comments
-  // static propTypes = {
-  //   comments: PropTypes.arrayOf(commentShape)
-  // };
+  static propTypes = {
+    comments: PropTypes.arrayOf(commentShape).isRequired,
+    token: PropTypes.string.isRequired
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      readyForReview: false,
+      comments: this.props.comments
```

Comment:
In general we want to avoid copying props into state because it can cause issues (for example, in the future if another component wants to use this and thinks it can control the comments via props it won't work). We should either have an onChange method that is handled by `TopInstructions` or put the comments in redux.

Reply:
> in the future if another component wants to use this and thinks it can control the comments via props it won't work).

For my understanding, can you clarify what you mean by this?

Reply:
if we have some component `SampleComponent` which has `ReviewTab` in it, and `SampleComponent` changes the `comments` props, `ReviewTab` will carry on with its copied state instead of using the new props (which is what we want here). This is an [article](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html) for a more up-to-date version of react, but generally copying props to state is an anti-pattern. LMK if that made sense

Reply:
Ok I think I'm getting it! It also mentions using a `key` as an option to force creating a new component instance (and all children) vs. updating an existing one, would that be ok, or is Redux preferable?

https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key

To test my understanding, am I correct that this component is in heavy violation of that anti-pattern?

https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/lib/levelbuilder/unit-editor/UnitEditor.jsx#L58

Reply:
I think either approach (key or redux) would be fine here. To my eyes that component does look in violation of the anti-pattern

Reply:
One question that comes to mind: What is the source of truth for comments that should be displayed? For example, if we add a refresh button that pulls in new comments from the DB, how will the new comments be synced with the existing comments?

If I'm understanding correctly, TopInstructions is handling information flow from the DB whereas ReviewTab is handling information flow from the "submit" box. With information being handled in two different places, we're reaching for the anti-pattern of storing props in state.

One possible solution:

TopInstructions seems to be doing too much. It's handling a _lot_ of display logic as well as managing a lot of information for each of its child components. Instead, we can have TopInstructions only decide whether or not to display the review tab.

The review tab, if displayed, would in turn handle the DB call as well as updates from the submit box. It would manage the source-of-truth list of comments.

The review tab would then continue to have child components that display the comments and the submit box.

Reply:
Two questions after looking at this briefly:
1. We remove `ReviewTab` from the DOM when we switch away from the tab, so if I move the AJAX call into the review tab (in, say, its own `componentDidMount`, we make a new request each time you switch to the tab). Maybe this is a positive from a refresh button perspective?
2. Just confirming that this feels worthwhile given how similar this feature is to teacher feedback to implement it in a different (even if better) way? May result in some context switching to understand we use different patterns to implement such closely related features. I think maybe the answer is yes (they're not entirely similar), but just want to make sure.

Reply:
Maybe easiest to just hide `ReviewTab` when it's not selected like we do here

https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/templates/instructions/TopInstructions.jsx#L700

Reply:
display: none seems like a frowned upon pattern as well :). Third idea (Molly's idea): store comments in Redux so we don't need to make network requests any time someone switches to the review tab.

Reply:
> display: none seems like a frowned upon pattern as well

^ TIL :) What's the reasoning for that?

Reply:
> so if I move the AJAX call into the review tab (in, say, its own componentDidMount, we make a new request each time you switch to the tab). Maybe this is a positive from a refresh button perspective?

^ That's a good point. If you don't mind, I'm going to schedule a message to Hannah and Jamila in the AM to see if that would be preferable.

Reply:
> Just confirming that this feels worthwhile given how similar this feature is to teacher feedback to implement it in a different (even if better) way?

^ I'd say so. A possible mitigation is to leave a comment indicating which pattern people should use going forward. That may also give a future engineer an easy blueprint to follow if they want to do a little bit of refactoring of TopInstructions.

Reply:
One final possibility (redux is also a great idea): [LibraryManagerDialog](https://github.com/code-dot-org/code-dot-org/blob/d86cd51e40578bf02853cd377503e7c3507f49bb/apps/src/code-studio/components/libraries/LibraryManagerDialog.jsx#L156) ran into a similar issue. The pattern followed there went like this:

LibraryManagerDialog mounts on page load but doesn't display anything.

Its parent indicates to LibraryManagerDialog whether it should be "open" or "closed"

When LibraryManagerDialog opens, it displays child components such as libraryListItem and triggers an ajax query. When the query is finished, LibraryManagerDialog updates its children to display the result of the ajax query. It also updates its own internal state.

When LibraryManagerDialog closes, it unmounts all of its child components, so nothing is displayed anymore. But LibraryManagerDialog itself is still mounted, so the state is preserved. Next time it is opened, the dialog is populated with the state maintained from the previous ajax query. The query will run again to take any updates from the server, but the user can start interacting with the UI immediately.

This may be too similar to the `display: none` scenario you mentioned earlier - definitely worth understanding whether this also violates the same principles.

Reply:
re: avoiding display: none, maybe totally frowned upon was a bit of an exaggeration, but I was just reading opinions about how you can pollute the DOM with too much hidden content (maybe not a huge problem in this case).

That said, I think using display: none here would result in the review tab being in the DOM everywhere that we show TopInstructions, which does feel like a code smell?

I have a couple of new commits that I'll share that moves both AJAX requests into `ReviewTab`. It is much cleaner I think, but does re-request on each click of the Review tab, which I can try and fix via Redux or the Library option you suggested, or do as a follow-up.